ToolsPanel: Refine component behaviour

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/34257

Depends on:
- https://github.com/WordPress/gutenberg/pull/34632

## Description

- Updates the `ToolsPanel` to improve its UX following the [suggestions outlined](https://github.com/WordPress/gutenberg/issues/34257#issuecomment-908554545) on the corresponding issue.
- Adds min-width to `ToolsPanel` dropdown menu
- Adds aria-labels to menu items

## How has this been tested?

Manually:

* Tested using the group block with padding and margin support
* Tested using changes from [34065](https://github.com/WordPress/gutenberg/pull/34065) and its injected min-height control for the cover block
* Manually updated storybook story to test various configurations

#### Test instructions
1. For testing, update the Group block's config to opt into padding and margin support, enabling one as a default control. 
```
		"spacing": {
			"padding": true,
			"margin": true,
			"__experimentalDefaultControls": {
				"padding": true
			}
		},
```
2. Depending on your theme, you'll likely also need to enable the margin support UI by setting the `settings.spacing.customMargin` property to `true` in your theme.json file.
3. Add a group block in the post editor and select it
4. Explore the behaviour of the ToolsPanel via the block's "Dimensions" panel.
5. Fire up native and confirm that the `Stepper` component displays its check icon correctly. Any `UnitControl` on native will show this `Stepper` component, one easy block to test with on native is the Cover block.

## Screenshots <!-- if applicable -->

#### Cover Block with cherry-picked commit from #34065
https://user-images.githubusercontent.com/60436221/132173157-a41ff448-2b83-43bb-ad54-877839013e25.mp4

#### Storybook
 https://user-images.githubusercontent.com/60436221/132362438-1a9ac63d-841d-4e62-ada9-d80b06b51e1e.mp4

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Separate default and optional controls in menu

Default and optional controls will now be displayed via separate menu groups and behave slightly differently. Default controls will always be shown and display a checkmark when empty or a minus icon when customized.

aaronrobertshaw committed:
Set min-width on ToolsPanel dropdown menu

aaronrobertshaw committed:
Make reset all a tertiary button as per design

aaronrobertshaw committed:
Add aria-labels to tools panel menu items

aaronrobertshaw commented:
It's worth noting here that the current iteration actually sets `disabled={ true }` on a default control's menu item when it does not have a customized value. In the mockup, the text for the item is normal and only the icon greyed out. Clicking the item would do nothing.  I think disabled in this context with the separation of default and optional controls makes sense.

Open to other ideas or opinions though.

jasmussen commented:
Thanks for tackling this one!

I had a bit of a challenge getting things to work as your recording; my Cover block still has a separate min-height panel, and I couldn't get margin to show despite trying a few different themes, and updating the Group block.json file. However I did see this:

![hey](https://user-images.githubusercontent.com/1204802/132187412-3a92a479-eb59-4ae7-891f-12a2aab78991.gif)

Some good steps here: the default panel is separate, I can reset it using the minus, I can reset all, and the check clears and removes non-default options.

We need the new icons from the mockup, these:

<img width="387" alt="Screenshot 2021-09-06 at 10 51 41" src="https://user-images.githubusercontent.com/1204802/132189144-d3541584-8af7-4622-b4a0-e7f68d32f7b2.png">

Updated `check`:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M16.7 7.1l-6.3 8.5-3.3-2.5-.9 1.2 4.5 3.4L17.9 8z" /></svg>
```

New `reset`:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7 11.5h10V13H7z" /></svg>
```

The new check is actually a bit smaller than the existing one:

<img width="366" alt="Screenshot 2021-09-06 at 10 55 18" src="https://user-images.githubusercontent.com/1204802/132189608-b572c092-aefe-47c4-abe0-8b911786088e.png">

I see that we already have a minus, I had forgotten that. Do you know if/where it's used at the moment? We might want to rename it to "reset" and replace it, depending on where and how much it's used.

I'm not sure whether we want a new `check-small` icon, or whether we just replace the existing one. I'm tempted to do the former. 

Let me know if you'd like me to add these separately.

ciampo commented on packages/components/src/tools-panel/tools-panel-header/component.js:
```diff
@@ -27,35 +41,87 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
+				<DropdownMenu
+					icon={ moreVertical }
+					label={ labelText }
+					menuProps={ { style: { minWidth: '200px' } } }
```
Ideally we'd like to avoid applying CSS via the inline `style` prop — the preferred option would be to pass them via a `className`

aaronrobertshaw replied:
I was under the impression I needed to avoid adding a stylesheet for components and didn't see how I could use styled-components within the `DropdownMenu` here. I'll revisit this as I work through the other suggestions. Thanks.

ciampo replied:
You could define styles in the `styles.js` file, and create a className by using the `cx` function in the hook.js file

<details>

<summary>Something like this (expand):</summary>

```diff
diff --git a/packages/components/src/tools-panel/styles.js b/packages/components/src/tools-panel/styles.js
index d9c7a985c0..c4b644fe9c 100644
--- a/packages/components/src/tools-panel/styles.js
+++ b/packages/components/src/tools-panel/styles.js
@@ -46,6 +46,10 @@ export const ToolsPanelHeader = css`
 	}
 `;
 
+export const DropdownMenu = css`
+	min-width: 200px;
+`;
+
 export const ToolsPanelItem = css`
 	grid-column: span 2;
 
diff --git a/packages/components/src/tools-panel/tools-panel-header/component.js b/packages/components/src/tools-panel/tools-panel-header/component.js
index 73af46f2be..5ff8304318 100644
--- a/packages/components/src/tools-panel/tools-panel-header/component.js
+++ b/packages/components/src/tools-panel/tools-panel-header/component.js
@@ -34,6 +34,7 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
 		menuItems,
 		resetAll,
 		toggleItem,
+		dropdownMenuClassName,
 		...headerProps
 	} = useToolsPanelHeader( props );
 
@@ -51,7 +52,7 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
 				<DropdownMenu
 					icon={ moreVertical }
 					label={ labelText }
-					menuProps={ { style: { minWidth: '200px' } } }
+					menuProps={ { className: dropdownMenuClassName } }
 				>
 					{ ( { onClose } ) => (
 						<>
diff --git a/packages/components/src/tools-panel/tools-panel-header/hook.js b/packages/components/src/tools-panel/tools-panel-header/hook.js
index c903c24725..73f7fe19fe 100644
--- a/packages/components/src/tools-panel/tools-panel-header/hook.js
+++ b/packages/components/src/tools-panel/tools-panel-header/hook.js
@@ -22,6 +22,10 @@ export function useToolsPanelHeader( props ) {
 		return cx( styles.ToolsPanelHeader, className );
 	}, [ className ] );
 
+	const dropdownMenuClassName = useMemo( () => {
+		return cx( styles.DropdownMenu );
+	}, [] );
+
 	const { menuItems, hasMenuItems } = useToolsPanelContext();
 
 	return {
@@ -29,5 +33,6 @@ export function useToolsPanelHeader( props ) {
 		hasMenuItems,
 		menuItems,
 		className: classes,
+		dropdownMenuClassName,
 	};
 }
```
</details>

aaronrobertshaw replied:
Brilliant, thanks! That certainly makes it obvious how it all fits together. 🙇 

I'll add those suggestions tomorrow.

aaronrobertshaw replied:
✅ Done.

ciampo commented on packages/components/src/tools-panel/tools-panel-header/component.js:
```diff
@@ -27,35 +41,87 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
-									}
-								) }
-							</MenuGroup>
+							{ !! defaultItems?.length && (
```
This block is starting to get quite hard to parse, have you considered extracting it to a separate component? (it could live in this same file and it wouldn't need to be exported).

Also, have you considered [using ternaries instead of the `&&` operator](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx)?

aaronrobertshaw replied:
> This block is starting to get quite hard to parse, have you considered extracting it to a separate component? 

Agreed. I've wavered on extracting the `MenuItem`s to their own components previously but it didn't reduce the code much. I'll look at refactoring the menu groups out into separate components instead.

> Also, have you considered using ternaries instead of the && operator?

In this instance, the use of `!!` was to specifically guard against the issue highlighted by that link. The `&&` operator still appears to be the prevailing style within the codebase. I understand that using a ternary makes missing the `!!` a non-issue. Is this something that should be added to the Gutenberg coding guidelines?

ciampo replied:
>  I'll look at refactoring the menu groups out into separate components instead.

Sounds good! Thank you :)

> In this instance, the use of !! was to specifically guard against the issue highlighted by that link. The && operator still appears to be the prevailing style within the codebase. I understand that using a ternary makes missing the !! a non-issue. Is this something that should be added to the Gutenberg coding guidelines?

Thank you for the explanation about using `!!`. I'm not sure if this should be added to the coding guidelines or even as an ESLint rule — that is something that @gziolo may advise about!

ciampo commented on packages/components/src/tools-panel/tools-panel-header/component.js:
```diff
@@ -27,35 +41,87 @@ const ToolsPanelHeader = ( props, forwardedRef ) => {
+												? minus
+												: check;
+
+											const itemLabel = hasValue
+												? sprintf(
+														// translators: %s: The name of the control being reset e.g. "Padding".
+														__( 'Reset %s' ),
+														label
+												  )
+												: undefined;
```
Nit: should this piece of logic also be moved to the `getAriaLabel` function? It seems weird to use that function only for optional items

aaronrobertshaw replied:
It does but so did adding extra conditionals within `getAriaLabel` along with passing extra args. In the end, I went with the simplest version of both the function and its use. I don't feel strongly about it, so happy to change as requested.

ciampo commented on packages/components/src/tools-panel/tools-panel/hook.js:
```diff
@@ -93,10 +119,11 @@ export function useToolsPanel( props ) {
 		// Turn off display of all non-default items.
-		const resetMenuItems = {};
+		const resetMenuItems = { default: {}, optional: {} };
 
 		panelItems.forEach( ( { label, isShownByDefault } ) => {
-			resetMenuItems[ label ] = !! isShownByDefault;
+			const group = isShownByDefault ? 'default' : 'optional';
+			resetMenuItems[ group ][ label ] = false;
```
Small nit: since this same logic is repeated earlier in this same file, I wonder if it could be extracted to improve the code's resistance to future changes (even by just creating a function private to this file)

ciampo reviewed (commented):
Thank you for starting to work on this, @aaronrobertshaw !

I've left a few initial inline comments, and I'm going to follow the progress of this PR in the upcoming days

ciampo commented:
>I had a bit of a challenge getting things to work as your recording; my Cover block still has a separate min-height panel, and I couldn't get margin to show despite trying a few different themes, and updating the Group block.json file

Just wanted to flag that I also had the same issues when trying to follow the instructions in the PR.

I also tried to use Storybook, but visiting the ToolsPanel story at the moment generates an error:

<img width="1792" alt="Screenshot 2021-09-07 at 10 35 19" src="https://user-images.githubusercontent.com/1083581/132317288-386647c5-2dec-4caf-bee0-6854230413f2.png">

aaronrobertshaw committed:
Fix bug where menuItems is undefined on initial render

aaronrobertshaw committed:
Tweak display add default controls to story

Adds default controls so that the updated behaviour separating default and optional controls into menu groups can be seen via the story.

aaronrobertshaw committed:
Apply min-width via CSS class

aaronrobertshaw committed:
Refactor default and optional control menu groups

aaronrobertshaw committed:
Move shared menu items generation logic to helper function

aaronrobertshaw committed:
Add new check and reset icons

aaronrobertshaw commented:
Thanks @jasmussen for your time and effort as always 👍 

> I had a bit of a challenge getting things to work as your recording; my Cover block still has a separate min-height panel, and I couldn't get margin to show despite trying a few different themes, and updating the Group block.json file. However I did see this:

The theme I was testing with already had the `customMargin` setting turned on enabling the UI. I believe this was the key step missing, so I've updated the PR test instructions accordingly.

To test with the cover block's min-height control injected via the proposed inspector controls slots, you'd need to apply commits from https://github.com/WordPress/gutenberg/pull/34065 on top of this PR. I did this to test the updated ToolsPanel with a SlotFill injected control, I didn't intend for you to need to do it at this point in time.

I fixed the issue with the ToolsPanel story that @ciampo highlighted so that is the easiest option for getting your hands on the updated panel now.

> We need the new icons from the mockup

Awesome, thank you. I've added the new icons.

> I see that we already have a minus, I had forgotten that. Do you know if/where it's used at the moment? We might want to rename it to "reset" and replace it, depending on where and how much it's used.

At a quick glance, it only appears that the minus icon is used for the iOS stepper.

> I'm not sure whether we want a new check-small icon, or whether we just replace the existing one. I'm tempted to do the former.

I added the icon as a new check icon rather than replace the existing one.

aaronrobertshaw commented:
I'm happy to update the check and reset icons however needed. Including handling them via a separate PR if the extra changes to existing icons are preferred.

jasmussen commented:
Alright, I promised I'd return with more info on the checks/reset icons:

- The check is not used widely enough that it's worth keeping a larger version. So let's replace the existing `check` with the new smaller vectors, instead of adding a new one.
- The existing `minus` icon was a Dashicon, and not part of the core Figma file. So let's replace the old vectors with the new (remember to update the viewBox from "-2 -2 24 24" to "0 0 24 24"), and rename it from `minus` to `reset`. This likely requires a breaking change note for the icon component.

I've updated the core Figma file with the new icons:

<img width="380" alt="Screenshot 2021-09-08 at 12 25 25" src="https://user-images.githubusercontent.com/1204802/132494815-2555c68c-df73-4b29-8554-597534c538ed.png">

Let me know if that works!

ciampo reviewed (commented):
From a quick look, code changes LGTM! I also tried the component in Storybook and it seemed to work as intended.

I'll let @jasmussen give the final seal of approval ✅

apeatling commented:
This is definitely an improvement, and should be merged. I do want to note that I think there's still too much assumed knowledge wrapped up in how this works. For example this first experience for someone who has never seen this before:

![2021-09-13 11 41 27](https://user-images.githubusercontent.com/1464705/133139128-8018fecd-7d16-4276-9402-96e80c436ec1.gif)

I don't think it's clear what the difference between the minus and checkbox really means. It's assuming I know that the padding is a default control and therefore can't be "unchecked". It's a tough one when only icons are being used with no labels indicating which controls are the default and can't be removed. I don't have any good suggestions, but wanted to note this.

apeatling reviewed (commented):
This is definitely an improvement, and should be merged. 

I do want to note that I think there's still too much assumed knowledge wrapped up in how this works. For example this first experience for someone who has never seen this before:

![2021-09-13 11 41 27](https://user-images.githubusercontent.com/1464705/133139128-8018fecd-7d16-4276-9402-96e80c436ec1.gif)

I don't think it's clear what the difference in expected action between the minus and check icon is. It's assuming I know that the padding is a default control and therefore can't be "unchecked". 

It's a tough one when only icons are being used with no labels indicating which controls are the default and therefore can't be removed. I don't have any good suggestions, but wanted to note this as something to think about further.

aaronrobertshaw committed:
Replace existing check icon

Instead of adding a second smaller check icon, we are replacing the existing check icon as it is not widely used.

aaronrobertshaw committed:
Replace minus icon with the new reset icon

aaronrobertshaw commented:
✅ The `check` and `minus` (now `reset`) icons have now been replaced with the supplied vectors
✅ Note about breaking change for `minus` icon added to changelog
✅ Confirmed native `Stepper` component still works after switching to `reset` icon
✅ Rebased this on top of trunk

aaronrobertshaw commented:
> I do want to note that I think there's still too much assumed knowledge wrapped up in how this works.

I agree with this. Solving it is a tricky one though.

> It's a tough one when only icons are being used with no labels indicating which controls are the default and therefore can't be removed. I don't have any good suggestions, but wanted to note this as something to think about further.

Adding some tooltips to the panel's menu items was trialled but it made the UI incredibly busy, undermining any improvements to the UX.

> This is definitely an improvement, and should be merged.

I'd be happy for this to land once @jasmussen has a chance to approve the PR in its current state. 

We can then iterate on this further in follow up PRs. Merging this will also allow for the `ToolsPanel` conversion to TypeScript to continue with less conflicts.

aaronrobertshaw committed:
Update color palette snapshots to use new check icon

jasmussen commented:
Took it for a quick spin, and the paddings, margins, and behavior looks right to me:

![tools](https://user-images.githubusercontent.com/1204802/133214093-fa7d195e-2950-40db-a0ee-cff3b48e2191.gif)

> I do want to note that I think there's still too much assumed knowledge wrapped up in how this works. 

The goal of the interface is to provide additional tools first and foremost, and secondarily to provide an additional and intentionally de-emphasized affordance for clearing values. Knowing that this interface can and will be iterated upon in the future, it feels important to start as simple as possible.

In that vein, let's definitely ship this one. And if we find the need to adjust further, we will. Thanks Aaron!

aaronrobertshaw committed:
Update color palette snapshots - take 2
